### PR TITLE
Feat/local image (WIP)

### LINKED ISSUE #5: Support downloading images to local & transformation with sharp
If `gatsby-source-flamelink` downloads remote images to local files, users can include images when building static sites, take advantage of Gatsby's responsive image solution such as [gatsby-image](https://www.gatsbyjs.org/packages/gatsby-image/), and reduce their Firebase's storage usage.

After a brief discussion with @jperasmus in the [slack channel](https://flamelink.io/slack), we drew out a basic plan:
- download all assets, either thru flamelink sdk with [`app.storage.getFiles`](https://flamelink.github.io/flamelink/#/storage?id=getfiles) or [`gsutils`](https://cloud.google.com/storage/docs/gsutil/commands/cp)
- cache it with Gatsby
- attach it to image fields

This issue will be used to track the development of this feature!

### PR DESCRIPTION
**Pull request by d4rekanguok**
Close #5

Hi @jperasmus, sorry for hogging this issue for so long! Shortly after our first discussion my deadline & priorities were all changed, so I ended up creating an ad-hoc solution on userland & rolled with it.

Recently @PatrickResponsive contacted me & show interested in working on this issue, so I quickly (& shamelessly) copied my dirty code over & create this PR to have something to discuss. I tried to at least finish the image locator function, so this PR is not so half-baked. I think my solution is not so great though, would love feedback from you guys.

Here's a rundown:

- Instead of downloading the whole media folders, I'm only downloading the image users use. For example, if users defined a 'hero' image field in their schema, only the images for that field of each post will be downloaded. 
- I couldn't figure out a smart way to find the image field, so I [recursively search each entry](https://github.com/flamelink/gatsby-source-flamelink/blob/301e5d5e405b998d69f29ad42bbfd541d8bf19ec/src/helpers/images.js#L10) for a field of this shape:
```
{
  contentType: 'image/****',
  url: 'https://....'
}
```
- Then [download & cache the image](https://github.com/flamelink/gatsby-source-flamelink/commit/f9d84ce6e4a82cfc89c9cdb168263ce9691a9d4f#diff-a8f14216da76798c0c02bbde9861df0bR34). I'm pretty much lifting the code from [`gatsby-source-wordpress`](https://github.com/gatsbyjs/gatsby/blob/9198af94d3b52933cb31c927cb267b1f37730f65/packages/gatsby-source-wordpress/src/normalize.js#L471) here, except for one difference:
  - We don't seem to keep different versions of the same image, so I don't check for last modified or created date to invalidate the cache.

- The downloaded image will be given a file node id, thanks to `gatsby-source-filesystem`. Then, I add this file node id as a `localFile___NODE` field to the image field (directly mutating it). Gatsby will recognizes the suffix `___NODE` & resolve the field `localFile` to its fileNode in graphql.

Then, users with a media field in their schema can query their images like so:
image

```
query Blog {
  flamelinkBlogContent {
    hero {
      localFile {
        name
        childImageSharp {
          id
        }
      }
    }
  }
}
```

and get this data:
```
{
  "data": {
    "flamelinkBlogContent": {
      "hero": [
        {
          "localFile": {
            "name": "flamelink%2Fmedia%2F1555643636571_01",
            "childImageSharp": {
              "id": "b6e23db4-b292-50ee-bd8f-8173ba6b4ff5"
            }
          }
        }
      ]
    }
  }
}
```

Things that I think should be worked on:

- Remove the recursive functions with something smarter, perhaps like looking for images based on the schema?
- Test with more use cases -- I've only test it with a file upload field in Flamelink's schema
- Make it so that the plugin can locate images in post body as well

I set up a Flamelink project to try it out. Perhaps I can share the test site & credential for it in Slack?

I also set up Jest to quickly test the image locator function, but I'm by no mean married to it, we can remove it / set up a different test method / whatever.

Sorry again about the delay!

### CODE REVIEW

**Comment by d4rekanguok** [comment]
@PatrickResponsive also I've just invited you to my fork so you can push stuff here if it makes sense to you!

**Comment by PatrickResponsive** [comment]
Awesome work thanks @d4rekanguok ,
Im gonna clone this and try get it working locally, Yeah, Ill try tackle refactoring this recursion function 👍

**Comment by jperasmus** [comment]
What we could do is use the schemas data and find all fields for each specific schema where the `type` is `media` and only use those.

File: src/helpers/images.js [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+ */
+const locateImageFields = entry => {
+  if (typeof entry !== 'object' || entry == null) return []
+
+  let results = []
+
+  // perfect condition
+  const { contentType, url } = entry
+  if (contentType && url) {
+    const ext = contentType.split('/')[1]
```
Comment (jperasmus) [inline src/helpers/images.js]: @d4rekanguok What is your thoughts around check for the specific `supportedExtensions` as the last part of the content type versus checking the first part, ie. checking for any content type that starts with `image/`?

Comment (d4rekanguok) [reply src/helpers/images.js]: At the time I thought it made sense to only download image type that `sharp` supports. But now come to think about it, perhaps it's better to download any `media` or `image` files locally, and then let `gatsby-transformer-sharp` decide whether to transform the file or not. We won't need to maintain a list of support extensions then!

Comment (jperasmus) [reply src/helpers/images.js]: Yes, I think that is a good idea. Then if the `gatsby-transformer-sharp` plugin rolls out more support, we get it by default.

Comment (d4rekanguok) [reply src/helpers/images.js]: I've quickly modified to fix this

**Comment by d4rekanguok** [comment]
@jp @PatrickResponsive Extending on JP's comment earlier, I tried to include images in a bunch of different places in a flamelink schema & got these 3 so far:
- a media field (`media`)
- a repeater field with nested media field (`repeater`)
- a relational select field (`select-relational`), that link to another schema with media field

We can check on these 3 type of fields in any schema to see if they includes images & go from there. (I've added a schema data in test folder in prepare for this.) I think we might run into image duplication problem when `populate` flag is enabled. For example, if I already have a image field in `tag` schema, then have a relational field in `blog` schema that points to `tag`, it may download the image in tag again.

**Comment by jperasmus** [comment]
Hi @d4rekanguok & @PatrickResponsive 

Apologies for my radio silence, been a bit busy on other work. Derek, I have not probably said thanks for this PR, so here goes: "thanks 👍". I love that you took this on in your spare time and made this plugin better.

Regarding the conflicting image ID's, what about prepending the schema's key and the specific field's key in front of the image ID, and do this like a breadcrumb as deep into the nested relationships as possible?

**Comment by d4rekanguok** [comment]
@jperasmus I'm happy to contribute!

The image duplication I mentioned earlier turns out to be a non-issue. Since the flamelink ID for each image is already unique, gatsby already knows enough to store only 1 file per image.

What I was worrying about was that gatsby downloads the same image multiple times on first run, since the download operations happens asynchronously:

Screen Shot 2019-05-28 at 4 50 10 PM

But it looks like `createRemoteFileNode` has [already taken care of this](https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby-source-filesystem/src/create-remote-file-node.js#L183).

**Comment by jperasmus** [comment]
ah awesome, that is very interesting. Thanks for the info. I am happy for this to be merged then. We can address any other issues as they come up.

**Comment by jperasmus** [comment]
This is now available in `v1.2.0`. Well done @d4rekanguok

**Comment by d4rekanguok** [comment]
awesome, thank you @jperasmus!

**Pull request merged**